File: backend/app/schemas/agents.py
from typing import Any

from pydantic import BaseModel, Field


class AgentInfo(BaseModel):
    id: int
    name: str
    role: str
    state: str
    description: str | None = None
    is_agent: bool = True
    agent_active: bool = True
    capability_ids: list[int] = Field(default_factory=list)
    max_concurrent_tasks: int
    open_assignment_count: int
    completed_assignment_count: int
    success_rate: float
    is_available: bool
    last_heartbeat: str | None = None
    discuss_channel_id: list[Any] | int | bool | None = None


class AgentHeartbeat(BaseModel):
    state: str | None = Field(default=None, pattern="^(idle|busy|blocked|offline|disabled)$")
    message: str | None = None


class AgentHeartbeatResponse(BaseModel):
    agent_id: int
    status: str
    message: str


class AgentAssignmentCreate(BaseModel):
    agent_id: int
    notes: str | None = None


class AgentAssignmentResponse(BaseModel):
    assignment_id: int
    task_id: int
    agent_id: int
    status: str
    created: bool


class AgentAuditLog(BaseModel):
    id: int
    agent_id: list[Any] | int | bool | None = None
    task_id: list[Any] | int | bool | None = None
    assignment_id: list[Any] | int | bool | None = None
    action: str
    message: str | None = None
    payload: dict[str, Any] | None = None
    created_at: str | None = None


class AgentRunRequest(BaseModel):
    limit: int = Field(default=1, ge=1, le=10)


class AgentChatMessage(BaseModel):
    message: str = Field(min_length=1, max_length=4000)


class AgentChatResponse(BaseModel):
    agent_id: int
    agent_name: str
    role: str
    reply: str


class AgentMessageCreate(BaseModel):
    body: str = Field(min_length=1, max_length=4000)
    task_id: int | None = None
    assignment_id: int | None = None


class AgentMessageResponse(BaseModel):
    agent_id: int
    status: str
    message: str


class AgentDiscussMessage(BaseModel):
    id: int
    body: str | None = None
    date: str | None = None
    author_id: list[Any] | int | bool | None = None


class AgentExecution(BaseModel):
    assignment_id: int
    agent_id: int
    task_id: int
    status: str
    progress: int
    summary: str
    command: str | None = None
    command_allowed: bool | None = None

File: frontend/src/lib/agents.ts
import { api } from "@/lib/api";

export type AgentInfo = {
  id: number;
  name: string;
  role: string;
  state: string;
  description: string | null;
  capability_ids: number[];
  max_concurrent_tasks: number;
  open_assignment_count: number;
  completed_assignment_count: number;
  success_rate: number;
  is_available: boolean;
  last_heartbeat: string | null;
};

export type AgentAuditLog = {
  id: number;
  agent_id: [number, string] | number | false | null;
  task_id: [number, string] | number | false | null;
  assignment_id: [number, string] | number | false | null;
  action: string;
  message: string | null;
  created_at: string | null;
};

export type AgentAssignment = {
  assignment_id: number;
  task_id: number;
  agent_id: number;
  status: string;
  created: boolean;
};

export type AgentExecution = {
  assignment_id: number;
  agent_id: number;
  task_id: number;
  status: string;
  progress: number;
  summary: string;
  command: string | null;
  command_allowed: boolean | null;
};

export async function fetchAgents() {
  const response = await api.get<AgentInfo[]>("/api/v1/agents");
  return response.data;
}

export async function fetchAgentAudit() {
  const response = await api.get<AgentAuditLog[]>("/api/v1/agents/audit");
  return response.data;
}

export async function recordAgentHeartbeat(agentId: number) {
  const response = await api.post(`/api/v1/agents/${agentId}/heartbeat`, {
    state: "idle",
    message: "Heartbeat from Arcvo Admin.",
  });
  return response.data;
}

export async function assignAgentTask(payload: { agent_id: number; task_id: number; notes: string }) {
  const response = await api.post<AgentAssignment>(
    `/api/v1/agents/tasks/${payload.task_id}/assign`,
    {
      agent_id: payload.agent_id,
      notes: payload.notes,
    },
  );
  return response.data;
}

export async function runAgent(agentId: number) {
  const response = await api.post<AgentExecution[]>(`/api/v1/agents/${agentId}/run`, {
    limit: 1,
  });
  return response.data;
}

export async function runPendingAgents() {
  const response = await api.post<AgentExecution[]>("/api/v1/agents/run-pending", {
    limit: 5,
  });
  return response.data;
}

export async function fetchAgentExecutions() {
  const response = await api.get<AgentAuditLog[]>("/api/v1/agents/executions");
  return response.data;
}

export type AgentChatResponse = {
  agent_id: number;
  agent_name: string;
  role: string;
  reply: string;
};

export async function chatWithAgent(agentId: number, message: string) {
  const response = await api.post<AgentChatResponse>(`/api/v1/agents/${agentId}/chat`, {
    message,
  });
  return response.data;
}
